designs/Transitive-Dependencies.md

```diff
@@ -0,0 +1,150 @@
+
+**By default, a user’s transitive packages will be collapsed** as a project can have hundreds, if not thousands of transitive packages & would potentially impact performance displaying these many records on each load.
+
+When a user highlights a transitive package, they will see a pop-up that displays how the transitive dependency originated & what top-level package(s) are bringing it in.
+
+![](../resources/TransitiveDependencies/TransitiveVSPMUI.png)
+
+Finally, selecting a transitive dependency will provide the user the ability to “Install” the package as you normally would a top-level dependency. This will promote the transitive dependency to a top-level dependency, and the transitive dependency will now be placed in the top-level packages list instead.
+
+## Goals and Non-Goals
+
+### Goals
+
+-	View transitive dependency list displaying the version of each dependency.
+-	See what brought in a transitive dependency & what version. 
+-	Ability to promote a transitive dependency to a top-level dependency via typical NuGet PM UI commands.
+
+### Non-Goals
+
+-	Complete management of transitive dependencies.
+-	Reinvent common UX patterns used in Visual Studio today for package management actions.
+
+## Design
+
+### Elements
+
+There are three primary elements that will be added as part of this design to the existing Visual Studio Package Manager UI.
+
+1.	ScrollViewer – To show transitive dependency nodes.
+2.	Expander – To collapse transitive dependency nodes by default & expand upon user action. To collapse top-level dependency nodes.
+3.	Tooltip – To show the transitive dependency’s origin.
+
+## Customer Development
+
+### Pain Points
+
+-	Frustration with lack of transparency for transitive packages during installs (did not realize installing package X would bring in so much other stuff)
+-	Frustration with lack of transparency for debugging reasons on PackageReference (customers feel they would be able to better debug versioning issues if they had a clearer view of their dependency graph)
+-	Some complaints about licensing issues in the case where a top-level package has a compatible license, but a transitive package does not.
+
+### Verbatims
+
+- Auditing/upgrading transitive dependencies. I can only do it manually by adding explicit packages.
+- Drilling down the dependency tree
+- In the package detail view, some way to explore dependencies of dependencies and then their dependencies and so on, any level deep, without installing the package.
+
+
+## Cross-cutting Concerns
+
+-	The words “top-level” & “transitive” are words most people must stop & think about.
+    -	Should we use “direct” & “indirect”? Or “explicit” & “implicit”?
```

Comment:
My preference would be "direct dependencies" and "indirect dependencies" here, if we're going for more simple language. I think this is descriptive enough without being too esoteric.

Reply:
I've seen the argument both directions, particularly top level vs transitive and direct vs indirect and in the end I think consistency matters.

My suggestion would be to double down to what we've already committed in top level and transitive. 

Note that other package managers such as paket, pub, npm already use top-level/direct & transitive. 

I've rarely seen indirect being used.

Reply:
UX Board will review this, but I have been a broken record of using language everyone already knows and is documented.

https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files for example uses top-level & transitive but it also uses direct. 🕺

Reply:
https://docs.microsoft.com/en-us/nuget/what-is-nuget also uses top-level & down-level

Reply:
`down-level`
🤕

Reply:
We are keeping the language top-level / transitive to be consistent. If UX board believes we should change it at a craftsman review, then we can reconsider.